File: 2018/python/day05.py
from string import ascii_lowercase


def react(polymer, skip):
    reacted, last = [], " "
    for char in polymer:
        if char.lower() == skip:
            continue
        if char != last and char.lower() == last.lower():
            del reacted[-1]
            try:
                last = reacted[-1]
            except:
                last = " "
            continue
        reacted.append(char)
        last = char
    return reacted


file = open("data/input-05.txt")
[polymer] = [i.strip() for i in file.readlines() if len(i) > 1]

res = len(react(polymer, " "))
print(f"Part 1: {res}")
assert res == 9296

res = min([len(react(polymer, c)) for c in ascii_lowercase])
print(f"Part 2: {res}")
assert res == 5534

File: 2018/julia/day05.jl
function react(polymer, skip)
    reacted, last = [], " "
    for char in polymer
        if lowercase(char) == skip
            continue
        end
        if char != last && lowercase(char) == lowercase(last)
            deleteat!(reacted, length(reacted))
            last = length(reacted) > 0 ? reacted[length(reacted)] : " "
            continue
        end
        push!(reacted, char)
        last = char
    end
    reacted
end

polymer = readlines("data/input-05.txt")[1]

res = length(react(polymer, " "))
println("Part 1: $(res)")
@assert(res == 9296)

res = minimum([length(react(polymer, c)) for c in 'a':'z'])
println("Part 2: $(res)")
@assert(res == 5534)
